SonarQube.Client/RoslynExporterAdapter/SonarLintConfigGenerator.cs

```diff
@@ -0,0 +1,108 @@
+
+namespace SonarQube.Client.RoslynExporterAdapter
+{
+    internal static class SonarLintConfigGenerator
+    {
+        private const string CSharpRepoKey = "csharpsquid";
+        private const string VBRepoKey = "vbnet";
+
+        public static SonarLintConfiguration Generate(IEnumerable<SonarQubeRule> rules, IDictionary<string, string> sonarProperties,
+            string language)
+        {
+            if (rules == null) { throw new ArgumentNullException(nameof(rules)); }
+            if (sonarProperties == null) { throw new ArgumentNullException(nameof(sonarProperties)); }
+            if (language == null) { throw new ArgumentNullException(nameof(language)); }
+
+            var slvsSettings = GetSettingsForLanguage(language, sonarProperties);
+
+            // We don't expect third-party rules to look for their settings in SonarLint.xml so we
+            // only fetch parameters for SonarC#/VB rules.
+            var sonarRepoKey = GetSonarRepoKey(language);
+            var slvsRules = GetRulesForRepo(sonarRepoKey, rules);
+
+            return new SonarLintConfiguration()
+            {
+                Settings = slvsSettings,
+                Rules = slvsRules
+            };
+        }
+
+        private static string GetSonarRepoKey(string language)
+        {
+            if (language == SonarQubeLanguage.CSharp.Key)
+            {
+                return CSharpRepoKey;
+            }
+
+            if (language == SonarQubeLanguage.VbNet.Key)
+            {
+                return VBRepoKey;
+            }
+
+            throw new ArgumentOutOfRangeException(nameof(language));
+        }
+
+        private static List<SonarLintKeyValuePair> GetSettingsForLanguage(string language, IDictionary<string, string> sonarProperties) =>
+            sonarProperties.Where(kvp => IsSettingForLanguage(language, kvp.Key))
+                .Select(ToSonarLintKeyValue)
+                .ToList();
+
+        private static bool IsSettingForLanguage (string language, string propertyKey)
+        {
+            var prefix = $"sonar.{language}.";
+
+            return propertyKey.StartsWith(prefix) &&
+                propertyKey.Length > prefix.Length;
+        }
+
+        private static List<SonarLintRule> GetRulesForRepo(string sonarRepoKey, IEnumerable<SonarQubeRule> sqRules) =>
+            sqRules.Where(ar => sonarRepoKey.Equals(ar.RepositoryKey))
+                .Select(ToSonarLintRule)
+                .ToList();
+
+        private static SonarLintRule ToSonarLintRule(SonarQubeRule sqRule)
+        {
+            List<SonarLintKeyValuePair> slvsParameters = null;
+            if (sqRule.Parameters != null && sqRule.Parameters.Count > 0 )
+            {
+                slvsParameters = sqRule.Parameters.Select(ToSonarLintKeyValue).ToList();
+            }
+
+            return new SonarLintRule()
+            {
+                Key = sqRule.Key,
```

Comment:
There is a reference to "TemplateKey" in the original, is that not needed?
https://github.com/SonarSource/sonar-scanner-msbuild/blob/9ccfdb648a0411014b29c7aee8e347aeab87ea71/src/SonarScanner.MSBuild.PreProcessor/Roslyn/RoslynSonarLint.cs#L58

Reply:
It doesn't exist as a field in the `SonarQubeRule` data class used by the client library, so it's never been used by SonarLint.

FYI I think the TemplateKey relates to "templated rules" - a simple but limited way of allowing users to add their custom rules. However, they are not supported by all languages (including C#, VB and C++).